File: simulator/backend/db/repositories/simulation-repository.ts
/**
 * Simulation Repository
 * Data access layer for simulation session operations
 */

import { sql } from '../connection.ts';
import type { SimulationData } from '@/shared/types/api.types';

export interface SimulationSession extends SimulationData {
  client_profile?: any;
  conversation_history?: any;
  review?: any;
  objectives?: string[];
}

export class SimulationRepository {
  /**
   * Find simulation by UUID
   */
  async findById(id: string): Promise<SimulationSession | null> {
    try {
      const result = await sql`
        SELECT id, simulation_id, user_id, industry, subcategory, difficulty, client_profile,
               conversation_history, objectives_completed, started_at, completed_at,
               total_xp, performance_review, duration_seconds
        FROM simulations
        WHERE id = ${id}
      `;
      return result[0] || null;
    } catch (error) {
      console.error('Error finding simulation by UUID:', error);
      throw error;
    }
  }

  /**
   * Find simulation by simulation_id (text identifier like "SIM-12345")
   */
  async findBySimulationId(simulationId: string): Promise<SimulationSession | null> {
    try {
      const result = await sql`
        SELECT id, simulation_id, user_id, industry, subcategory, difficulty, client_profile,
               conversation_history, objectives_completed, started_at, completed_at,
               total_xp, performance_review, duration_seconds
        FROM simulations
        WHERE simulation_id = ${simulationId}
      `;
      return result[0] || null;
    } catch (error) {
      console.error('Error finding simulation by simulation_id:', error);
      throw error;
    }
  }

  /**
   * Get simulations by user ID
   */
  async findByUserId(userId: string): Promise<SimulationSession[]> {
    try {
      const result = await sql`
        SELECT id, simulation_id, user_id, industry, subcategory, difficulty,
               started_at, completed_at, total_xp, performance_review
        FROM simulations
        WHERE user_id = ${userId}
        ORDER BY started_at DESC
      `;
      return result;
    } catch (error) {
      console.error('Error fetching simulations by user ID:', error);
      throw error;
    }
  }

  /**
   * Get all simulations
   */
  async findAll(limit?: number): Promise<SimulationSession[]> {
    try {
      const query = limit
        ? sql`SELECT id, simulation_id, user_id, industry, subcategory, difficulty,
               started_at, completed_at, total_xp, performance_review
             FROM simulations
             ORDER BY started_at DESC
             LIMIT ${limit}`
        : sql`SELECT id, simulation_id, user_id, industry, subcategory, difficulty,
               started_at, completed_at, total_xp, performance_review
             FROM simulations
             ORDER BY started_at DESC`;

      return await query;
    } catch (error) {
      console.error('Error fetching all simulations:', error);
      throw error;
    }
  }

  /**
   * Create simulation session
   */
  async create(sessionData: {
    user_id: string;
    industry: string;
    difficulty: string;
    subcategory?: string;
    simulation_id?: string;
    client_profile?: any;
    objectives_completed?: any[];
  }): Promise<SimulationSession> {
    try {
      const result = await sql`
        INSERT INTO simulations (
          simulation_id, user_id, industry, subcategory, difficulty,
          client_profile, objectives_completed, started_at
        )
        VALUES (
          ${sessionData.simulation_id || `SIM-${Date.now()}`},
          ${sessionData.user_id},
          ${sessionData.industry},
          ${sessionData.subcategory || null},
          ${sessionData.difficulty},
          ${sessionData.client_profile ? JSON.stringify(sessionData.client_profile) : '{}'},
          ${sessionData.objectives_completed ? JSON.stringify(sessionData.objectives_completed) : '[]'},
          NOW()
        )
        RETURNING id, simulation_id, user_id, industry, subcategory, difficulty,
                  client_profile, conversation_history, objectives_completed,
                  started_at, completed_at, total_xp, performance_review, duration_seconds
      `;
      return result[0];
    } catch (error) {
      console.error('Error creating simulation session:', error);
      throw error;
    }
  }

  /**
   * Update simulation session
   */
  async update(id: string, updates: {
    conversation_history?: any;
    completed_at?: Date;
    total_xp?: number;
    performance_review?: any;
    duration_seconds?: number;
  }): Promise<SimulationSession> {
    try {
      const updateFields: string[] = [];

      if (updates.conversation_history !== undefined) {
        updateFields.push(`conversation_history = '${JSON.stringify(updates.conversation_history)}'`);
      }
      if (updates.completed_at !== undefined) {
        updateFields.push(`completed_at = '${updates.completed_at.toISOString()}'`);
      }
      if (updates.total_xp !== undefined) {
        updateFields.push(`total_xp = ${updates.total_xp}`);
      }
      if (updates.performance_review !== undefined) {
        updateFields.push(`performance_review = '${JSON.stringify(updates.performance_review)}'`);
      }
      if (updates.duration_seconds !== undefined) {
        updateFields.push(`duration_seconds = ${updates.duration_seconds}`);
      }

      if (updateFields.length === 0) {
        throw new Error('No fields to update');
      }

      const result = await sql`
        UPDATE simulations
        SET ${sql.raw(updateFields.join(', '))}
        WHERE id = ${id}
        RETURNING id, simulation_id, user_id, industry, subcategory, difficulty,
                  client_profile, conversation_history, objectives_completed,
                  started_at, completed_at, total_xp, performance_review, duration_seconds
      `;

      return result[0];
    } catch (error) {
      console.error('Error updating simulation session:', error);
      throw error;
    }
  }

  /**
   * Complete simulation
   */
  async complete(id: string, total_xp: number, performance_review: any): Promise<SimulationSession> {
    try {
      const result = await sql`
        UPDATE simulations
        SET completed_at = NOW(),
            total_xp = ${total_xp},
            performance_review = ${JSON.stringify(performance_review)}
        WHERE id = ${id}
        RETURNING id, simulation_id, user_id, industry, subcategory, difficulty,
                  started_at, completed_at, total_xp, performance_review
      `;
      return result[0];
    } catch (error) {
      console.error('Error completing simulation:', error);
      throw error;
    }
  }

  /**
   * Delete simulation
   */
  async delete(id: string): Promise<void> {
    try {
      await sql`
        DELETE FROM simulations
        WHERE id = ${id}
      `;
    } catch (error) {
      console.error('Error deleting simulation:', error);
      throw error;
    }
  }

  /**
   * Get user statistics
   */
  async getUserStats(userId: string): Promise<{
    total: number;
    completed: number;
    avgScore: number;
  }> {
    try {
      const result = await sql`
        SELECT
          COUNT(*) as total,
          COUNT(completed_at) as completed,
          AVG(score) as avg_score
        FROM simulations
        WHERE user_id = ${userId}
      `;

      return {
        total: parseInt(result[0].total),
        completed: parseInt(result[0].completed),
        avgScore: parseFloat(result[0].avg_score) || 0,
      };
    } catch (error) {
      console.error('Error fetching user stats:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const simulationRepository = new SimulationRepository();

File: backend-python/app/repositories/simulation_repository.py
"""
Simulation Repository
Database operations for simulation entities
"""

import json
import logging
from typing import Any
from uuid import uuid4

from app.database import fetch, fetchone, execute, record_to_dict, records_to_list
from app.models.simulation import SimulationData

logger = logging.getLogger(__name__)


class SimulationRepository:
    """Repository for simulation database operations."""

    async def find_by_id(self, id: str) -> SimulationData | None:
        """Find simulation by UUID."""
        try:
            query = """
                SELECT id, simulation_id, user_id, industry, subcategory, difficulty, client_profile,
                       conversation_history, objectives_completed, total_xp, performance_review,
                       started_at, completed_at, duration_seconds
                FROM simulations WHERE id = $1
            """
            record = await fetchone(query, id)
            if not record:
                return None
            data = record_to_dict(record)
            # Parse JSON fields
            for field in ["client_profile", "conversation_history", "objectives_completed", "performance_review"]:
                if data.get(field) and isinstance(data[field], str):
                    try:
                        data[field] = json.loads(data[field])
                    except json.JSONDecodeError:
                        pass
            return SimulationData(**data)
        except Exception as e:
            logger.error(f"Error finding simulation by ID: {e}")
            raise

    async def find_by_simulation_id(self, simulation_id: str) -> SimulationData | None:
        """Find simulation by simulation_id (text identifier like SIM-12345)."""
        try:
            query = """
                SELECT id, simulation_id, user_id, industry, subcategory, difficulty, client_profile,
                       conversation_history, objectives_completed, total_xp, performance_review,
                       started_at, completed_at, duration_seconds
                FROM simulations WHERE simulation_id = $1
            """
            record = await fetchone(query, simulation_id)
            if not record:
                return None
            data = record_to_dict(record)
            # Parse JSON fields
            for field in ["client_profile", "conversation_history", "objectives_completed", "performance_review"]:
                if data.get(field) and isinstance(data[field], str):
                    try:
                        data[field] = json.loads(data[field])
                    except json.JSONDecodeError:
                        pass
            return SimulationData(**data)
        except Exception as e:
            logger.error(f"Error finding simulation by simulation_id: {e}")
            raise

    async def find_by_user_id(self, user_id: str) -> list[SimulationData]:
        """Find all simulations for a user."""
        try:
            query = """
                SELECT id, simulation_id, user_id, industry, subcategory, difficulty, client_profile,
                       conversation_history, objectives_completed, total_xp, performance_review,
                       started_at, completed_at, duration_seconds
                FROM simulations WHERE user_id = $1
                ORDER BY started_at DESC
            """
            records = await fetch(query, user_id)
            simulations = []
            for record in records:
                data = record_to_dict(record)
                # Parse JSON fields
                for field in ["client_profile", "conversation_history", "objectives_completed", "performance_review"]:
                    if data.get(field) and isinstance(data[field], str):
                        try:
                            data[field] = json.loads(data[field])
                        except json.JSONDecodeError:
                            pass
                simulations.append(SimulationData(**data))
            return simulations
        except Exception as e:
            logger.error(f"Error finding simulations by user ID: {e}")
            raise

    async def find_all(self, limit: int | None = None) -> list[SimulationData]:
        """Get all simulations."""
        try:
            query = """
                SELECT id, simulation_id, user_id, industry, subcategory, difficulty, client_profile,
                       conversation_history, objectives_completed, total_xp, performance_review,
                       started_at, completed_at, duration_seconds
                FROM simulations
                ORDER BY started_at DESC
            """
            if limit:
                query += f" LIMIT {limit}"
            records = await fetch(query)
            simulations = []
            for record in records:
                data = record_to_dict(record)
                for field in ["client_profile", "conversation_history", "objectives_completed", "performance_review"]:
                    if data.get(field) and isinstance(data[field], str):
                        try:
                            data[field] = json.loads(data[field])
                        except json.JSONDecodeError:
                            pass
                simulations.append(SimulationData(**data))
            return simulations
        except Exception as e:
            logger.error(f"Error fetching all simulations: {e}")
            raise

    async def create(self, simulation_data: dict[str, Any]) -> SimulationData:
        """Create a new simulation."""
        try:
            # Generate simulation_id text identifier (e.g., SIM-12345678)
            import time
            sim_id_text = simulation_data.get("simulation_id") or f"SIM-{int(time.time() * 1000) % 100000000}"

            query = """
                INSERT INTO simulations (
                    simulation_id, user_id, industry, subcategory, difficulty, client_profile,
                    conversation_history, objectives_completed, started_at
                )
                VALUES ($1, $2, $3, $4, $5, $6, $7, $8, NOW())
                RETURNING id, simulation_id, user_id, industry, subcategory, difficulty, client_profile,
                          conversation_history, objectives_completed, total_xp, performance_review,
                          started_at, completed_at, duration_seconds
            """
            record = await fetchone(
                query,
                sim_id_text,
                simulation_data["user_id"],
                simulation_data["industry"],
                simulation_data.get("subcategory"),
                simulation_data.get("difficulty", "beginner"),
                json.dumps(simulation_data.get("client_profile", {})),
                json.dumps(simulation_data.get("conversation_history", [])),
                json.dumps(simulation_data.get("objectives_completed", [])),
            )
            data = record_to_dict(record)
            for field in ["client_profile", "conversation_history", "objectives_completed", "performance_review"]:
                if data.get(field) and isinstance(data[field], str):
                    try:
                        data[field] = json.loads(data[field])
                    except json.JSONDecodeError:
                        pass
            return SimulationData(**data)
        except Exception as e:
            logger.error(f"Error creating simulation: {e}")
            raise

    async def update(self, simulation_id: str, update_data: dict[str, Any]) -> SimulationData:
        """Update a simulation."""
        try:
            # Build dynamic update query
            update_parts = []
            params = [simulation_id]
            param_idx = 2

            for key, value in update_data.items():
                if value is not None:
                    # Serialize JSON fields
                    if key in ["client_profile", "conversation_history", "objectives_completed", "performance_review"]:
                        value = json.dumps(value) if not isinstance(value, str) else value
                    update_parts.append(f"{key} = ${param_idx}")
                    params.append(value)
                    param_idx += 1

            if not update_parts:
                return await self.find_by_id(simulation_id)

            query = f"""
                UPDATE simulations SET {', '.join(update_parts)}
                WHERE id = $1
                RETURNING id, simulation_id, user_id, industry, subcategory, difficulty, client_profile,
                          conversation_history, objectives_completed, total_xp, performance_review,
                          started_at, completed_at, duration_seconds
            """
            record = await fetchone(query, *params)
            if not record:
                raise ValueError("Simulation not found")
            data = record_to_dict(record)
            for field in ["client_profile", "conversation_history", "objectives_completed", "performance_review"]:
                if data.get(field) and isinstance(data[field], str):
                    try:
                        data[field] = json.loads(data[field])
                    except json.JSONDecodeError:
                        pass
            return SimulationData(**data)
        except Exception as e:
            logger.error(f"Error updating simulation: {e}")
            raise

    async def complete(
        self,
        simulation_id: str,
        total_xp: int,
        performance_review: dict[str, Any] | None = None,
    ) -> SimulationData:
        """Mark simulation as complete."""
        try:
            query = """
                UPDATE simulations
                SET completed_at = NOW(), total_xp = $2, performance_review = $3
                WHERE id = $1
                RETURNING id, simulation_id, user_id, industry, subcategory, difficulty, client_profile,
                          conversation_history, objectives_completed, total_xp, performance_review,
                          started_at, completed_at, duration_seconds
            """
            record = await fetchone(
                query,
                simulation_id,
                total_xp,
                json.dumps(performance_review) if performance_review else None,
            )
            if not record:
                raise ValueError("Simulation not found")
            data = record_to_dict(record)
            for field in ["client_profile", "conversation_history", "objectives_completed", "performance_review"]:
                if data.get(field) and isinstance(data[field], str):
                    try:
                        data[field] = json.loads(data[field])
                    except json.JSONDecodeError:
                        pass
            return SimulationData(**data)
        except Exception as e:
            logger.error(f"Error completing simulation: {e}")
            raise

    async def delete(self, simulation_id: str) -> None:
        """Delete a simulation."""
        try:
            query = "DELETE FROM simulations WHERE id = $1"
            await execute(query, simulation_id)
        except Exception as e:
            logger.error(f"Error deleting simulation: {e}")
            raise

    async def get_user_stats(self, user_id: str) -> dict[str, Any]:
        """Get user simulation statistics."""
        try:
            query = """
                SELECT
                    COUNT(*) as total,
                    COUNT(completed_at) as completed,
                    COALESCE(AVG(total_xp), 0) as avg_score
                FROM simulations WHERE user_id = $1
            """
            record = await fetchone(query, user_id)
            if not record:
                return {"total": 0, "completed": 0, "avgScore": 0}
            return {
                "total": record["total"],
                "completed": record["completed"],
                "avgScore": float(record["avg_score"]) if record["avg_score"] else 0,
            }
        except Exception as e:
            logger.error(f"Error getting user stats: {e}")
            raise


# Singleton instance
simulation_repository = SimulationRepository()
